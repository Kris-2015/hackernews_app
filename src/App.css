.page {
    margin: 20px;
}
.interactions {
    text-align: center;
}

.card {
    top: 60px !important;
    left: 150px !important;
}