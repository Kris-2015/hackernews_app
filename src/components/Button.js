import React, { Component } from 'react';

/**
 * @class Button
 * @purpose Component which accepts some properties
 */
class Button extends Component
{
    render () {
        const {
            onClick,
            className,
            type,
            children,
        } = this.props;

        return (
            <button
                onClick={onClick}
                className={className}
                type={type}
            >
                {children}
            </button>
        );
    }
}

export default Button;