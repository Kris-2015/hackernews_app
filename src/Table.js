import React, { Component } from 'react';
import Button from './Button';
import { sortBy } from 'lodash'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import constants from './config/constants';
import Sort from './Sort';

/**
 * @class Table
 * @purpose Table component which displays data in grid
 */
class Table extends Component
{
    render() {
        // bootstrap class column; Destructing the constants with different name
        const {
            ID_COLUMN: idColumn,
            LARGE_COLUMN: largeColumn,
            MID_COLUMN: midColumn,
            SMALL_COLUMN:smallColumn,
        } = constants;

        const SORTS  = {
            NONE: list => list,
            TITLE: list => sortBy(list, 'title'),
            AUTHOR: list => sortBy(list, 'author'),
        };

        const {
            list,
            sortKey,
            isSortReverse,
            onSort,
            onDismiss,
        } = this.props;

        const sortedList = SORTS[sortKey] (list);
        const reverseSortedList = isSortReverse ?
            sortedList.reverse() : sortedList;

        return (
            <div className="table">
                <div className="table-header">
                    <span className={ idColumn }>
                        <p>ID</p>
                    </span>
                    <span className={ largeColumn }>
                        <Sort
                            sortKey={'TITLE'}
                            onSort={onSort}
                            activeSortKey = { sortKey }
                            >
                            <p>Title</p>
                        </Sort>
                    </span>
                    <span className={ midColumn }>
                        <Sort
                            sortKey={'AUTHOR'}
                            onSort={onSort}
                            activeSortKey = { sortKey }
                        ><p>Author</p>
                        </Sort>
                    </span>
                    <span className={ smallColumn }>
                        <Button className="button-inline">
                            <p>Remove</p>
                        </Button>
                    </span>
                </div>
                {reverseSortedList.map((item, index) =>
                    <div key={item.objectID} className="table-row">
                        <span className={ idColumn }>
                            <p>{index + 1}</p>
                        </span>
                        <span className={ largeColumn }>
                            <a href={item.url} target="_blank">{item.title}</a>
                        </span>
                        <span className={ midColumn }>{item.author}</span>
                        <span className={ smallColumn }>
                            <Button
                                onClick={() => onDismiss(item.objectID)}
                                className="btn btn-danger btn-size"
                            >
                            <i>
                                <FontAwesomeIcon icon="trash" />
                            </i>
                            </Button>
                        </span>
                    </div>
                )}
            </div>
        );
    };
}

export default Table;