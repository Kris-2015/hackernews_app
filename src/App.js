import React, { Component } from 'react';
import axios from 'axios';
import './App.css';
import Table from './Table';
import Button from "./Button";
import Search from './Search';
import config from './config/config';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import SlideToggle from './SlideToggle';
import Loading from './Loading';

/**
 * @class App
 * @purpose Core part of the application
 */
class App extends Component {

    /**
     * @param props
     * @purpose define state and bind class functions
     */
    constructor (props) {
        super(props);

        this.state = {
            results: null,
            searchKey: '',
            searchTerm: config.DEFAULT_QUERY,
            error: null,
            isLoading: false,
            sortKey: 'NONE',
            isSortReverse: false,
        };

        // It indicates whether the component has been mounted or not
        this._isMounted = false;
    }


    /**
     * Function to check whether api call is required
     * @param searchTerm
     * @return boolean
     */
    needToSearchTopStories = (searchTerm) => {
        // Get the result from cache
        let cacheHits = localStorage.getItem(searchTerm);
        let cachePage = localStorage.getItem('page');

        // Return result from cache if cache data is present
        if (cacheHits) {
            // Update the state of the result with cache data
            this.setState({
                results: {
                    [searchTerm] : { hits : JSON.parse(cacheHits),
                        page: JSON.parse(cachePage)
                    },
                },
                isLoading: false
            });

            console.log('returning data from cache');
            return false;
        }

        console.log('Make an api call to get data!');
        return true;
        // Commenting the line for future reference
        //return !this.state.results[searchTerm];
    };

    /**
     * Function to set search result in result state
     * @param result
     * @return void
     */
    setSearchTopStories = (result) => {
        // Destructing the state: reference ES6 Destructure
        const { hits, page } = result;
        const { searchKey, results, error } = this.state;

        // Return old when search key is present
        const oldHits = results && results[searchKey]
            ? results[searchKey].hits
            : [];

        // Update the hits data, if new data is present
        const updateHits = [
            ...oldHits,
            ...hits
        ];

        // Set the state of the error if it is set to true
        if (error) {
            this.setState({error: false});
        }

        // Append the current result with old result
        this.setState({
            results: {
                ...results,
                [searchKey] : { hits : updateHits, page },
            },
            isLoading: false
        }, () => {
            // Store the result and page limit in local storage by search key
            localStorage.setItem(searchKey, JSON.stringify(updateHits));
            localStorage.setItem('page', JSON.stringify(page));
        });
    };

    /**
     * Function to get the search term when search event is triggered
     * @param event
     * @return void
     */
    onSearchChange = (event) => this.setState({ searchTerm: event.target.value });

    /**
     * Function to fetch stories from hacker-news api
     * @param searchTerm
     * @param page
     * @return void
     */
    fetchSearchTopStories = (searchTerm, page=0) => {
        // Set the state of the isLoading to true, when new search request is made
        this.setState({ isLoading: true });

        // Make api calls when network is present
        if (navigator.onLine) {
            axios(`${config.PATH_BASE}${config.PATH_SEARCH}?${config.PARAM_SEARCH}${searchTerm}&${config.PARAM_PAGE}\
${page}&${config.PARAM_HPP}${config.DEFAULT_HPP}`)
                .then(result => this._isMounted && this.setSearchTopStories(result.data))
                .catch(error => this._isMounted && this.setState({ error }));
        } else {
            // Set the error state to true if application is offline
            this.setState({ error: true});
        }
    };

    /**
     * React lifecycle component which gets called after render function
     * @purpose After component has been mounted, perform search operation
     * @return void
     */
    componentDidMount() {
        this._isMounted = true;

        const { searchTerm } = this.state;
        this.setState({ searchKey: searchTerm });

        // Check whether to make api call or not
        if (this.needToSearchTopStories(searchTerm)) {
            this.fetchSearchTopStories(searchTerm);
        }
    }

    /**
     * React lifecycle component which will be called after component has mounted
     * @purpose Remove the declared event listener and set isMounted flag to false
     * @return void
     */
    componentWillUnmount() {
        this._isMounted = false;
    }

    /**
     * Function to perform search operation when search request has been triggered
     * @purpose Get the requested search term and perform search operation
     * @return void
     */
    onSearchSubmit = (event) => {
        const { searchTerm } = this.state;
        this.setState({ searchKey: searchTerm });

        // Make api call when network connection is alive
        if (this.needToSearchTopStories(searchTerm)) {
            this.fetchSearchTopStories(searchTerm);
        }

        event.preventDefault();
    };

    /**
     * Function to remove the row when clicks trash button
     * @purpose Update the result of current search key
     * @return void
     */
    onDismiss = (id) => {
        const { searchKey, results } = this.state;
        const { hits, page } = results[searchKey];
        const isNotId = item => item.objectID !== id;

        // Remove the particular row from the hit list i.e, search results
        const updatedHits = hits.filter(isNotId);

        // Store the updated result
        localStorage.setItem(searchKey, JSON.stringify(updatedHits));

        this.setState({
            results: {
                ...results,
                [searchKey]: { hits: updatedHits, page }
            }
        });
    };

    /**
     * Function to sort the search results
     * @purose Perform the sorting based on search key
     * @return void
     */
    onSort = (sortKey) => {
        const isSortReverse = this.state.sortKey === sortKey &&
            !this.state.isSortReverse;

        this.setState({ sortKey, isSortReverse });
    };

    render() {
        const {
            searchTerm,
            searchKey,
            results,
            error,
            isLoading,
            sortKey,
            isSortReverse
        } = this.state;

        const page = (
            results &&
            results[searchKey] &&
            results[searchKey].page
        ) || 0;

        const list = (
            results &&
            results[searchKey] &&
            results[searchKey].hits
        ) || [];

        return (
            <div className="page">

                {/* Append the component of react-detect-offline*/}
                <SlideToggle />

                <div className="interactions">
                    {/* Search Component */}
                    <Search
                        value={ searchTerm }
                        onChange={ this.onSearchChange }
                        onSubmit={ this.onSearchSubmit }
                    >
                        <i><FontAwesomeIcon icon="search" /></i>
                    </Search>
                </div>
                {/*Applying Conditional Rendering in ternary operation for result*/}
                { error
                    ?<div className="interactions">
                        <p>Oops, Something went wrong!</p>
                    </div>
                    :
                    <Table
                        list={ list }
                        sortKey = { sortKey }
                        isSortReverse = { isSortReverse }
                        onSort = { this.onSort }
                        onDismiss = { this.onDismiss }
                    />
                }

                {/* More button to generate more result as a paginated data */}
                <div className="interactions">
                    {isLoading
                        ? <Loading />
                        :
                        <Button
                            onClick={() => this.fetchSearchTopStories(searchTerm, page + 1)}
                            className="btn btn-primary"
                            type="button"
                        >
                            More
                        </Button>
                    }
                </div>
            </div>
        );
    }
}

export default App;