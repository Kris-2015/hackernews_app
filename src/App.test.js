import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import App, { Search } from './App';
import Button from './Button';
/*';
import Table from './Table';*/

describe('Button', () => {

    it('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(<Button/>, div);
        ReactDOM.unmountComponentAtNode(div);
    });

    test('has a valid snapshot', () => {
        const component = renderer.create(
            <Button/>
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

});
