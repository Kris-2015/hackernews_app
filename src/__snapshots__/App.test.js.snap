// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button has a valid snapshot 1`] = `
<button
  className={undefined}
  onClick={undefined}
  type="button"
/>
`;

exports[`Search has a valid snapshot 1`] = `
<form
  onSubmit={undefined}
>
  <input
    onChange={undefined}
    type="text"
    value={undefined}
  />
  <button
    className="btn btn-primary btn-size"
    type="submit"
  >
    Search
  </button>
</form>
`;
